src/PFire.Core/Protocol/Messages/Outbound/FriendsSessionAssign.cs

```diff
@@ -32,20 +32,13 @@ public override void Process(XFireClient client)
         {
             var friends = client.Server.Database.QueryFriends(_ownerUser);
 
-            // offline friends are just not sent
-            var friendsSessions = friends.Select(a => client.Server.GetSession(a))
-                                         .Where(a => a != null)
-                                         .ToList();
-
-            friendsSessions.ForEach(session =>
+            foreach (var friend in friends)
             {
-                UserIds.Add(session.User.UserId);
-                SessionIds.Add(session.SessionId);
-                Debug.WriteLine("Status: For:{0} -- FriendId:{1} Friend:{2} FriendSession:{3}", client.User.Username, session.User.UserId, session.User.Username, session.SessionId);
-            });
-            
-            //friendsSessions.OrderBy(a => a.User.UserId);
-            //friendsSessions.ForEach(session => Debug.WriteLine("Context: For:{0} -- FriendId:{1} Friend:{2} FriendSession:{3}", context.User.Username, session.User.UserId, session.User.Username, session.SessionId));
+                var friendSession = client.Server.GetSession(friend);
+
+                UserIds.Add(friend.UserId);
+                SessionIds.Add(friendSession != null ? friendSession.SessionId : Guid.Empty);
```

Comment:
We shouldn't be adding an empty Guid if a session isn't found, instead just don't add anything at all. Might be nice to implement a similar pattern on collections of `TryGet`. Then we can nicely only add to the `SessionIds` if it's true

Reply:
According to the docs "If the session ID is 0, this indicates that the friend has gone offline" and certainly with my two pcs here, exit the xfire client on one, with this code, the client on the first pc reports that the other client has gone off line.

Reply:
Oh wow - I didn't know that about the client. I'm curious how the client knows which friend has gone offline? I suppose that when it recieves a 0 session id, it does a diff and removes missing ones. Intresting - and great find!

I think to help make this intention clear for future readers we could set Guid.empty to a class level variable called `InvalidateFriendSession` perhaps?

Reply:
Sure.. good call.

The client gets a list of all it's accepted friends, and if they are online, a non-zero session id is provided with the userid. If the user is offline, an 0 session is provided.

Will do these updates, and push the branch again

Reply:
Thanks!